src/routes/notifications/v1/notifications.controller.ts

```diff
@@ -11,27 +12,119 @@ import { RegisterDeviceDto } from '@/routes/notifications/v1/entities/register-d
+      await this.createV2RegisterDto(registerDeviceDto);
+
+    for (const compatibleV2Request of compatibleV2Requests) {
+      await this.notificationServiceV2.upsertSubscriptions(compatibleV2Request);
+    }
+  }
+
+  private async createV2RegisterDto(args: RegisterDeviceDto): Promise<
+    Array<{
+      upsertSubscriptionsDto: UpsertSubscriptionsDto;
+      authPayload: AuthPayload;
+    }>
+  > {
+    const safeV2Array: Array<{
+      authPayload: AuthPayload;
+      upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+        signature: `0x${string}`;
+      };
+    }> = [];
+
+    const safesV1Registrations = args.safeRegistrations;
+
+    for (const safeV1Registration of safesV1Registrations) {
+      if (safeV1Registration.safes.length) {
+        const safeV2: {
+          authPayload: AuthPayload;
+          upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+            safes: Array<UpsertSubscriptionsSafesDto>;
+            signature: `0x${string}`;
+          };
+        } = {
+          upsertSubscriptionsDto: {
+            cloudMessagingToken: args.cloudMessagingToken,
+            deviceType: args.deviceType,
+            deviceUuid: (args.uuid as UUID) || undefined,
+            safes: [],
+            signature: safeV1Registration.signatures[0] as `0x${string}`,
+          },
+          authPayload: new AuthPayload(),
+        };
+        for (const safeAddresses of safeV1Registration.safes) {
+          safeV2.upsertSubscriptionsDto.safes.push({
+            address: safeAddresses as `0x${string}`,
+            chainId: safeV1Registration.chainId,
+            notificationTypes: Object.values(NotificationType),
+          });
+        }
+        safeV2Array.push(safeV2);
+      }
+    }
+
+    for (const [index, safeV2] of safeV2Array.entries()) {
+      const safeAddresses = safeV2.upsertSubscriptionsDto.safes.map(
+        (safeV2Safes) => safeV2Safes.address,
+      );
+
+      const recoveredAddress = await recoverMessageAddress({
+        message: `gnosis-safe${args.timestamp}${args.uuid}${args.cloudMessagingToken}${safeAddresses.sort().join('')}`,
+        signature: safeV2.upsertSubscriptionsDto.signature,
+      });
+
+      safeV2.authPayload.chain_id =
+        safeV2.upsertSubscriptionsDto.safes[0].chainId;
+      safeV2.authPayload.signer_address = recoveredAddress;
+
+      safeV2Array[index].authPayload = safeV2.authPayload;
+    }
+
+    return safeV2Array;
   }
 
   @Delete('chains/:chainId/notifications/devices/:uuid')
   async unregisterDevice(
     @Param('chainId') chainId: string,
     @Param('uuid') uuid: string,
   ): Promise<void> {
-    return this.notificationsService.unregisterDevice({ chainId, uuid });
+    await this.notificationsService.unregisterDevice({ chainId, uuid });
+
+    // Compatibility with V2
+    // @TODO Remove NotificationModuleV2 after all clients have migrated and compatibility is no longer needed.
+    await this.notificationServiceV2.deleteDevice(uuid as UUID);
```

Comment:
We could use the `UuidSchema` here and not need cast.

Reply:
Done
https://github.com/safe-global/safe-client-gateway/pull/2063/commits/4e45430b47c033acb313cd5f14162d3b81142b4b